src/Microsoft.VisualStudio.Threading/AsyncLazy`1.cs

```diff
@@ -72,6 +72,30 @@ public AsyncLazy(Func<Task<T>> valueFactory, JoinableTaskFactory? joinableTaskFa
         this.jobFactory = joinableTaskFactory;
     }
 
+    /// <summary>
+    /// Gets a value indicating whether to suppress detection of a value factory depending on itself.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// A value factory that truly depends on itself (e.g. by calling <see cref="GetValueAsync()"/> on the same instance)
+    /// would deadlock, and by default this class will throw an exception if it detects such a condition.
+    /// However this detection relies on the .NET ExecutionContext, which can flow to "spin off" contexts that are not awaited
+    /// by the factory, and thus could legally await the result of the value factory without deadlocking.
+    /// </para>
+    /// <para>
+    /// When this flows improperly, it can cause <see cref="InvalidOperationException"/> to be thrown, but only when the value factory
+    /// has not already been completed, leading to a difficult to reproduce race condition.
+    /// Such a case can be resolved by calling <see cref="SuppressRelevance"/> around the non-awaited fork in <see cref="ExecutionContext" />,
+    /// or the entire instance can be configured to suppress this check by setting this property to <see langword="true"/>.
+    /// </para>
+    /// <para>
+    /// When this property is set to <see langword="true" />, the recursive factory check will not be performed,
+    /// but <see cref="SuppressRelevance"/> will still call into <see cref="JoinableTaskContext.SuppressRelevance"/>
+    /// if a <see cref="JoinableTaskFactory"/> was provided to the constructor.
+    /// </para>
+    /// </remarks>
+    public bool SuppressRecursiveFactoryDetection { get; init; }
```

Comment:
It would be good to document the default value.

Reply:
Done.